logstash-core/lib/logstash/plugin_metadata.rb

```diff
@@ -0,0 +1,108 @@
+# encoding: utf-8
+
+require 'thread_safe/cache'
+
+module LogStash
+  ##
+  # `PluginMetadata` provides a space to store key/value metadata about a plugin, typically metadata about
+  # external resources that can be gleaned during plugin registration.
+  #
+  # Data is persisted across pipeline reloads, and no effort is made to clean up metadata from pipelines
+  # that no longer exist after a pipeline reload.
+  #
```

Comment:
I would think it potentially problematic if data were persisted across pipeline reloads if each plugin's data is keyed by the plugin's ID. It's quite common for plugin ID to be auto-generated GUIDs which change with each pipeline reload. That means that with each pipeline reload, the PluginMetadata structure would retain stale data about old plugins and users could have difficulty identifying which entries refer to the currently-running plugins.

Also, not clearing PluginMetadata across pipeline reloads could result in the reporting of incorrect data if a plugin with a statically-defined ID in an updated pipeline does not correctly reset its own metadata. That seems like something that Logstash should make impossible by clearing PluginMetadata when a pipeline is reloaded.